feat: add playback queue

andykelk opened the pull request:
## Summary

Adds a playback queue feature allowing users to line up items to play sequentially. The queue persists across sessions and is managed at both the JavaScript/Vue layer and the native Android layer for seamless background playback.

- **Queue management UI**: New `QueueModal` component with drag-to-reorder, play from position, and individual item removal. Accessible via a queue button in the audio player, side drawer, and item more menus
- **Add to queue**: Queue buttons added to item/episode more menus, podcast latest episode rows, and the episode detail page
- **Persistent storage**: Queue is saved to Capacitor `Preferences` (SharedPreferences on Android) and restored on app launch
- **Native Android auto-advance**: `QueueManager` reads the queue directly from SharedPreferences so the Android `PlayerNotificationService` can advance to the next item on playback completion, even when the app is backgrounded
- **Progress preservation**: Current playback position is saved in the queue item so playback resumes from where it left off
- **Native→JS sync**: When the native layer advances the queue, it emits `onQueueChanged` so the Vue store reloads from storage and stays in sync

## Changed files

- `store/index.js` — queue state, getters, mutations and actions
- `store/globals.js` — `showQueueModal` flag
- `plugins/localStore.js` — `getPlaybackQueue` / `setPlaybackQueue` methods
- `mixins/queueMixin.js` — shared `addItemToQueue` helper used across components
- `components/modals/QueueModal.vue` — queue management modal
- `components/app/AudioPlayer.vue` — queue button + `onQueueChanged` listener
- `components/app/SideDrawer.vue` — queue menu item
- `components/modals/ItemMoreMenuModal.vue` — "Add to Queue" action
- `components/tables/podcast/LatestEpisodeRow.vue` — queue button
- `pages/item/_id/_episode/index.vue` — "Add to Queue" action
- `layouts/default.vue` — mount `QueueModal`, load saved queue on init
- `android/.../data/PlaybackQueueItem.kt` — data model matching JS queue item structure
- `android/.../managers/QueueManager.kt` — reads/writes queue from SharedPreferences
- `android/.../media/MediaProgressSyncer.kt` — mark finished even when timer is stopped
- `android/.../player/PlayerNotificationService.kt` — native queue auto-advance on playback end
- `android/.../player/AbMediaDescriptionAdapter.kt` — `invalidateCache()` for cover art refresh
- `android/.../plugins/AbsAudioPlayer.kt` — `onQueueChanged` event bridge
- `strings/en-us.json` — `ButtonAddToQueue`, `ButtonClearQueue`, `LabelQueue`, `MessageQueueEmpty`

## Test plan

- [ ] Add an audiobook to the queue from the item more menu; verify it appears in the queue modal
- [ ] Add a podcast episode to the queue from the episode page and latest episodes list
- [ ] Drag items to reorder in the queue modal; verify order is persisted after closing and reopening
- [ ] Delete an item from the queue; verify it is removed
- [ ] Play an item from the queue modal while something is already playing; verify the current item is added back to the front of the queue
- [ ] Let playback finish; verify the next queue item starts automatically
- [ ] Force-close the app and reopen; verify the queue is restored from storage
- [ ] Test background auto-advance (screen off, app backgrounded) on Android

andykelk committed:
Add playback queue state management and strings

- Add playbackQueue array to Vuex store state
- Add queue getters: getPlaybackQueue, getNextQueueItem, hasQueueItems, getQueueLength
- Add queue actions: addToQueue, removeFromQueue, clearQueue, moveQueueItem, playNextInQueue
- Add queue mutations: addToQueue, removeFromQueue, clearQueue, moveQueueItem, setPlaybackQueue
- Add queue-related strings: LabelQueue, ButtonAddToQueue, ButtonClearQueue, MessageQueueEmpty

Implements core data management for issue #416 queue functionality.

andykelk committed:
Add QueueModal component for queue management

- Create draggable queue modal with item reordering
- Show queue items with cover art, title, author, and duration
- Support play, remove, and clear queue actions
- Empty state with appropriate messaging
- Integrates with vuedraggable for drag-and-drop reordering

Part of issue #416 queue functionality implementation.

andykelk committed:
fix: update close modal pattern

Use the same pattern as other modals for the close button for the queue
modal.

andykelk committed:
feat: add queue button for items and episodes

Using a mixin to contain the common logic, add to the "more" menu for
both items and episodes.

andykelk committed:
feat: allow queue access from the side drawer

If nothing is playing, we need a way to access the queue. Add this as an
item in the side drawer.

This uses the same modal as is available from the player view.

andykelk committed:
feat: use device storage to persist queue

Previous implementation lost the queue on app restart.

andykelk committed:
fix: use correct images for queue items

andykelk committed:
feat: preserve playback progress when adding items to queue

Episodes and audiobooks now resume from their previous position when played from queue instead of starting from the beginning.

andykelk committed:
fix: queue modal button functionality and drag interactions

- Fixed play and delete buttons not responding to clicks
- Resolved draggable component interfering with button interactions
- Added proper drag handles to separate draggable areas from buttons
- Made buttons always visible instead of hover-only for better mobile UX
- Fixed async ordering issues when playing items from queue

andykelk committed:
fix: improve Android background playback and auto-advance

Prevent PlayerNotificationService from stopping when app is removed from recent apps if audio is actively playing, allowing proper background playback and auto-advance functionality.

andykelk committed:
feat: add queue button to latest podcast episodes

Added "Add to Queue" button to the latest episode rows in podcast views, allowing users to easily queue episodes directly from the latest episodes list.

andykelk committed:
refactor: move queue menu item below local media in side drawer

Repositioned the queue menu item to appear after local media/downloads section for better logical grouping and improved navigation flow.

andykelk committed:
fix: standardize queue icons to playlist_play for consistency

Updated all queue-related icons from queue_music to playlist_play across the application to better distinguish queue functionality from playlists and indicate the immediate playback nature of queues.

andykelk committed:
feat: add PlaybackQueueItem data model for native queue support

- Create PlaybackQueueItem.kt matching JavaScript queue structure
- Add JSON serialization/deserialization utilities using Jackson
- Include helper methods for media ID matching and type checking
- Foundation for native queue management to fix screen-off auto-advance

This is Phase 1, Commit 1 of native queue implementation.
No functional changes - preparation only.

andykelk committed:
feat: add QueueManager to read queue from Capacitor storage

- Create QueueManager.kt to access queue from SharedPreferences
- Read queue data from CapacitorStorage (same storage as web layer)
- Provide methods: getQueue(), getNextItem(), hasQueueItems()
- Read-only implementation for now - no queue modifications
- Add debug logging for queue state inspection

This is Phase 1, Commit 2 of native queue implementation.
No functional changes yet - preparation only.

andykelk committed:
feat: wire up QueueManager to PlayerNotificationService

Initialize QueueManager in onCreate() to enable access to playback queue
from native Android layer. This completes the foundation phase with all
necessary infrastructure in place for implementing queue auto-advance.

andykelk committed:
feat: implement native queue auto-advance

Add queue checking logic to handlePlaybackEnded() to automatically play
the next item when playback ends. This enables queue auto-advance to work
even when the screen is off, as the native Android layer continues running
in the background.

Features:
- Check for queue items after playback ends
- Load library items from local DB or server API
- Handle both audiobooks and podcast episodes
- Resume from saved progress position
- Maintain user's playback speed

This solves the critical bug where auto-advance rarely works when the
phone screen is off, as JavaScript event loops are paused in that state.

andykelk committed:
feat: add queue item removal after playback starts

Add removeFirstItem() method to QueueManager to remove the first item
from the queue and persist the change to SharedPreferences. Call this
method in handlePlaybackEnded() after getting the next queue item but
before attempting to play it, following the same pattern as the web layer.

This ensures items are removed from the queue as they're played, preventing
the same item from being played multiple times.



fix: add missing import for toJsonArray extension function

andykelk committed:
feat: delegate queue auto-advance to native player layer

Move queue auto-advance logic from JavaScript to the native Android
player layer to ensure it works reliably when the screen is off and
JavaScript execution is paused.

The native PlayerNotificationService now:
- Reads the queue from SharedPreferences on playback end
- Automatically plays the next item if available
- Continues to work even when the device screen is locked

This ensures uninterrupted playback across queued items regardless
of screen state or JavaScript execution context.

andykelk committed:
feat: add native-to-JavaScript queue synchronization

Implement event-driven sync to keep JavaScript layer updated when the
native Android layer modifies the queue during screen-off playback:

- Add onQueueChanged() event to ClientEventEmitter interface
- Emit event from native layer after queue modification
- Handle event in JavaScript to reload queue from SharedPreferences
- Fix loadSavedQueue to update state even when queue is empty
- Remove queue items when playback starts, not when it ends
- Ensure queue change notification sent before playback session update

This ensures the queue UI stays in sync with the native player's
auto-advance behavior when the screen is off, including proper handling
of the last queue item.

andykelk committed:
fix: sync progress before queue auto-advance

Ensure queue items are marked as played on the server by calling
mediaProgressSyncer.finished() before auto-advancing to the next item.
Previously only the first manually-started item would be synced, while
auto-advanced items were not marked as finished.

The finished() method only executes its callback if listeningTimerRunning
is true, so we conditionally call it only when the timer is running.
Otherwise we proceed directly with queue advance. This ensures:
- Queue auto-advance always works (playNextItem function always executes)
- Progress is synced when possible (when timer is still running)

This matches the behavior of Android Auto podcast auto-advance which
correctly syncs progress before playing the next episode.

andykelk committed:
fix: mark queue items as finished even when timer stopped

The player automatically pauses when playback ends, which triggers
onIsPlayingChanged(false) and calls mediaProgressSyncer.pause(),
stopping the timer. This happened BEFORE handlePlaybackEnded() was
called, so finished() would return early without syncing.

Modified finished() to:
- Always execute the callback, even when timer already stopped
- Manually sync progress with full duration when needed
- Handle both server and local items correctly
- Save local progress and sync with server when connected

This ensures all queue items are properly marked as played on the
server regardless of when the progress syncer timer was stopped.

nichwall commented on android/app/src/main/java/com/audiobookshelf/app/player/AbMediaDescriptionAdapter.kt:
```diff
@@ -91,4 +91,9 @@ class AbMediaDescriptionAdapter (private val controller: MediaControllerCompat,
     }
   }
+
+  fun invalidateCache() {
```
This function is not referenced anywhere, is it intended to be used?

Atinoda commented:
Thank you @advplyr for Audiobookshelf, and @andykelk for this feature!

I was keen to get using this (currently listening to a podcast with frustratingly short episodes...) so I rebased the feature branch to master and built an apk. It is working well so far. Will report back with any bugs and some feedback after using it for a while. If there's anything you want me to look at in particular, please let me know.

andykelk committed:
fix: code review cleanup

- Remove redundant same-package imports from PlayerNotificationService.kt
- Simplify QueueModal.vue by removing unused handlePlayClick/handleDeleteClick wrappers
- Add missing newline at end of queueMixin.js
- Fix LabelQueue alphabetical ordering in en-us.json

Rapha149 commented:
Hi,
firstly I want to say that I really appreciate your work on this since a queue is the main thing I miss for podcasts on ABS.
However, I would like to suggest two changes:

1. Add a button for enqueuing an episode in the podcast overview page where the episodes of that podcast are listed, like in the latest page. 

2. For the button in the latest page: I find the popups that appear when adding episodes to the queue a bit annoying, especially when I enqueue multiple episodes, as the popups quickly take up half the screen.  
I would suggest this alternative solution: don't allow multiple instances of the same episode / audiobook in the queue, as in my opinion this wouldn't make much sense anyway for these media types. Then the button on the latest page could get an indicator whether the episode is currently enqueued or not, e.g. a green color when it is. The color of the button would be direct feedback and could replace the popup. If it is pressed again there are two options: either the episode is removed from the queue or the queue display opens.  
In the same manner the "Add to queue" button in the three dot menus could be replaced by "Remove from queue" when the episode or audiobook is already enqueued.

Anyway these are just suggestions, please let me know what you think!

andykelk committed:
Merge remote-tracking branch 'upstream/master' into feature/playback-queue

andykelk committed:
feat: improved UX for adding to queue

Add a button to add/remove to queue on the podcast overview page -
allows adding multiple episodes from one podcast in quick succession.

Switch the "add to queue" function to a toggle - if the item is in the
queue already, it gets removed; if it doesn't exist in the queue, it
gets added.

andykelk commented:
> Hi, firstly I want to say that I really appreciate your work on this since a queue is the main thing I miss for podcasts on ABS. However, I would like to suggest two changes:

Good ideas. I have incorporated these in the newest commit.

Rapha149 commented:
Wow, that was fast! Thank you, it looks very good.

One small thing I also noticed: there is no haptic feedback for any of the buttons. For example, the button for marking an episode as finished provides haptic feedback (in the latest page as well as the three dot menu). Adding this would be a very small change, if I'm not mistaken all you would have to do is make the methods `async` and add `await this.$hapticsImpact()`.

Atinoda commented:
Hi Andy, thank you for working on this!

My apologies that I never got back to you, but I will echo the comments of Rapha from my experience, and having just built and tested the app the fixes are lovely and work well.

- **Previous and Next track buttons**: These do not seem to work for queue navigation, but I think they did before. Maybe a regression?

- **Auto-enqueue podcast episodes as filtered and displayed**: Something to consider - perhaps a related PR is #1816. Way back when, this was the option / behaviour I expected from the app, and it would align to the web client behaviour. Then I discovered there is no queue (sad), then I discovered you had made one (happy). Along that journey, it seems I forgot where the first footstep was placed.

- **Progress / complete status flags and the sleep timer auto-rewind feature**: My use-case was sleepytime listening to shorter podcasts (multiple episodes and fiddling with the buttons at the end of each was annoying). My (admittedly bleary-eyed) testing seemed to have surfaced an edge case. If the rewind timer horizon goes back beyond the current episode, I don't think that progress and completion flags are are being managed. For example, queue of E1 and E2 both episodes have a 15m length; sleep timer of 25m then rewind timer of 20m; the playback sequence is `{E1 [complete], E2 [progress 10m]}` and the rewind timer should then reset this to `{E1 [incomplete, 5m progress], E2 [progress none]}` Honestly though - any time this is relevant I have literally fallen asleep so this is vaguely remembered! Given how fiddly that is to describe, the logic to address that might be torturous with the required queue trajectory memory and state tracking... so I'm just raising it to you but not necessarily suggesting it is something worth working on.

Thanks again for your work in developing this feature!

andykelk committed:
fix: queue navigation via next/previous buttons and haptic feedback

- Fix skipToNext() to advance the playback queue when pressing the
  next button (previously it only worked for multi-track audiobooks).
  Uses stop() instead of finished() so the current item is not marked
  as 100% complete when manually skipping.
- Extract startNextQueueItem() helper to deduplicate queue advance
  logic between handlePlaybackEnded() and skipToNext().
- Add missing haptic feedback to queue toggle buttons in EpisodeRow,
  LatestEpisodeRow, QueueModal (play/remove/clear), ItemMoreMenuModal,
  and the episode detail page.

andykelk committed:
fix: enable skip next/previous actions in media session PlaybackState

Without ACTION_SKIP_TO_NEXT and ACTION_SKIP_TO_PREVIOUS in the
PlaybackState, Android media controllers (lock screen, Bluetooth
headphones, car) grey out the buttons regardless of whether the
handler code is correct. Add the actions to both
getSupportedQueueNavigatorActions() and
setMediaSessionConnectorPlaybackActions() so they appear enabled.

andykelk committed:
fix: show next/previous buttons in notification player

Enable next/previous in the expanded media notification so queue
navigation is accessible. Keep them out of the compact view to
preserve the existing rewind/play/fast-forward compact layout.

andykelk committed:
fix: enable next button in player when queue has items

The in-app last_page button was always disabled for podcasts because it
only checked for a next chapter. Now it also enables when the playback
queue has items, and clicking it will advance to the next queue item
when there is no next chapter to jump to.
